src/main/java/com/icthh/xm/uaa/service/TenantPermissionService.java

```diff
@@ -39,16 +44,22 @@
     // Map structure: <Tenant, <Role, [Permission]>>
     final Map<String, Map<String, List<Permission>>> tenantRolePermissions = new ConcurrentHashMap<>();
 
-    public List<AccPermissionDTO> getEnabledPermissionByRole(String role) {
+    public List<AccPermissionDTO> getEnabledPermissionByRole(List<String> roles) {
 
         String tenant = TenantContextUtils.getRequiredTenantKeyValue(tenantContextHolder.getContext());
 
-        return tenantRolePermissions.getOrDefault(tenant, Collections.emptyMap())
-                                    .getOrDefault(role, Collections.emptyList())
-                                    .stream()
-                                    .filter(permission -> !permission.isDisabled())
-                                    .map(AccPermissionDTO::new)
-                                    .collect(Collectors.toList());
+        return ofNullable(roles).orElse(emptyList()).stream().map(role ->
+            tenantRolePermissions.getOrDefault(tenant, Collections.emptyMap())
+                .getOrDefault(role, emptyList())
+                .stream()
+                .filter(permission -> !permission.isDisabled())
+                .map(AccPermissionDTO::new)
+                .collect(Collectors.toList()))
+            .filter(CollectionUtils::isNotEmpty)
+            .flatMap(Collection::stream)
+            .collect(Collectors.toSet())
+            .stream()
+            .collect(Collectors.toList());
```

Comment:
@ibilous this code can be simplified as:
```
        public List<AccPermissionDTO> getEnabledPermissionByRole(List<String> roles) {

        Map<String, List<Permission>> tenantRoles = getTenantRoles();

        return ofNullable(roles)
            .orElseGet(Collections::emptyList)
            .stream()
            .map(tenantRoles::get)
            .filter(Objects::nonNull)
            .flatMap(Collection::stream)
            .filter(permission -> !permission.isDisabled())
            .map(AccPermissionDTO::new)
            .collect(Collectors.toList());
    }

    private Map<String, List<Permission>> getTenantRoles(){
        String tenant = TenantContextUtils.getRequiredTenantKeyValue(tenantContextHolder);
        return tenantRolePermissions.getOrDefault(tenant, Collections.emptyMap());
    }
 ````

Reply:
fixed